File: car_reminders.js
// car_reminders.js — JavaScript версия

const fs = require('fs');
const readline = require('readline');

const DATA_FILE = 'reminders.json';
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

let reminders = [];

function load() {
    if (fs.existsSync(DATA_FILE)) {
        try {
            reminders = JSON.parse(fs.readFileSync(DATA_FILE, 'utf8'));
        } catch (e) {
            reminders = [];
        }
    }
}

function save() {
    fs.writeFileSync(DATA_FILE, JSON.stringify(reminders, null, 2));
}

function ask(question) {
    return new Promise(resolve => rl.question(question, resolve));
}

function color(text, code) {
    return `\x1b[${code}m${text}\x1b[0m`;
}

async function main() {
    load();
    while (true) {
        console.log(`\n${color('🔔 Дневник напоминаний (JavaScript)', '36')}`);
        console.log("1. Добавить напоминание");
        console.log("2. Показать все напоминания");
        console.log("3. Показать предстоящие напоминания");
        console.log("4. Поиск напоминаний");
        console.log("5. Удалить напоминание");
        console.log("6. Отметить как выполненное");
        console.log("7. Статистика");
        console.log("8. Выход");
        const choice = await ask("Выберите действие: ");
        switch (choice.trim()) {
            case "1": await addReminder(); break;
            case "2": listAll(); break;
            case "3": await listUpcoming(); break;
            case "4": await searchReminders(); break;
            case "5": await deleteReminder(); break;
            case "6": await markDone(); break;
            case "7": stats(); break;
            case "8": console.log("До свидания!"); rl.close(); return;
            default: console.log(color("Неверный выбор.", "31"));
        }
    }
}

async function addReminder() {
    const title = await ask("Название: ");
    const date = await ask("Дата (ГГГГ-ММ-ДД): ");
    const mileage = parseInt(await ask("Пробег (км): "));
    const desc = await ask("Описание: ");
    let priority = (await ask("Приоритет (низкий/средний/высокий): ")).toLowerCase();
    if (!["низкий", "средний", "высокий"].includes(priority)) priority = "средний";
    const id = reminders.length + 1;
    reminders.push({ id, title, date, mileage, description: desc, priority, done: false });
    save();
    console.log(color(`✅ Напоминание добавлено (ID: ${id})`, "32"));
}

function listAll() {
    if (reminders.length === 0) {
        console.log(color("Нет напоминаний.", "33"));
        return;
    }
    console.log(color(`${'ID'.padEnd(4)} ${'Название'.padEnd(20)} ${'Дата'.padEnd(12)} ${'Пробег'.padEnd(10)} ${'Приоритет'.padEnd(10)} ${'Статус'.padEnd(12)}`, "36"));
    console.log("-".repeat(80));
    for (const r of reminders) {
        const status = r.done ? "✅ Выполнено" : "⏳ Ожидает";
        const priorityColor = r.priority === "низкий" ? "32" : r.priority === "средний" ? "33" : "31";
        console.log(`${String(r.id).padEnd(4)} ${r.title.padEnd(20)} ${r.date.padEnd(12)} ${String(r.mileage).padEnd(10)} ${color(r.priority.padEnd(10), priorityColor)} ${status}`);
    }
}

async function listUpcoming() {
    const days = parseInt(await ask("Количество дней (по умолч. 7): ")) || 7;
    const now = new Date();
    const upcoming = reminders.filter(r => {
        if (r.done) return false;
        const d = new Date(r.date);
        return d >= now && d <= new Date(now.getTime() + days * 24 * 60 * 60 * 1000);
    });
    if (upcoming.length === 0) {
        console.log(color(`Нет напоминаний на ближайшие ${days} дней.`, "32"));
        return;
    }
    console.log(color(`📅 Напоминания на ближайшие ${days} дней:`, "36"));
    for (const r of upcoming) {
        const diff = Math.ceil((new Date(r.date) - now) / (1000 * 60 * 60 * 24));
        console.log(`  ${r.id}: ${r.title} — ${r.date} (осталось ${diff} дн)`);
    }
}

async function searchReminders() {
    const keyword = (await ask("Введите ключевое слово: ")).toLowerCase();
    const results = reminders.filter(r => r.title.toLowerCase().includes(keyword) || r.description.toLowerCase().includes(keyword));
    if (results.length === 0) {
        console.log(color("Ничего не найдено.", "33"));
    } else {
        results.forEach(r => console.log(`${r.id}: ${r.title} | ${r.date} | ${r.priority} | ${r.done ? '✅' : '⏳'}`));
    }
}

async function deleteReminder() {
    listAll();
    const id = parseInt(await ask("Введите ID для удаления: "));
    const index = reminders.findIndex(r => r.id === id);
    if (index !== -1) {
        reminders.splice(index, 1);
        save();
        console.log(color("✅ Напоминание удалено.", "32"));
    } else {
        console.log(color("❌ Напоминание не найдено.", "31"));
    }
}

async function markDone() {
    listAll();
    const id = parseInt(await ask("Введите ID для отметки как выполненное: "));
    const r = reminders.find(r => r.id === id);
    if (r) {
        r.done = true;
        save();
        console.log(color("✅ Напоминание отмечено как выполненное.", "32"));
    } else {
        console.log(color("❌ Напоминание не найдено.", "31"));
    }
}

function stats() {
    const total = reminders.length;
    const done = reminders.filter(r => r.done).length;
    const pending = total - done;
    const now = new Date();
    const overdue = reminders.filter(r => !r.done && new Date(r.date) < now).length;
    console.log(color("📊 Статистика:", "36"));
    console.log(`  Всего напоминаний: ${total}`);
    console.log(`  Выполнено: ${done}`);
    console.log(`  Ожидает: ${pending}`);
    console.log(`  Просрочено: ${overdue}`);
}

main().catch(console.error);

File: car_reminders.cs
// car_reminders.cs — C# версия

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;

class Reminder
{
    public int Id { get; set; }
    public string Title { get; set; }
    public string Date { get; set; }
    public int Mileage { get; set; }
    public string Description { get; set; }
    public string Priority { get; set; }
    public bool Done { get; set; }
}

class Program
{
    private static List<Reminder> reminders = new List<Reminder>();
    private const string DataFile = "reminders.json";

    static void Main()
    {
        Load();
        while (true)
        {
            Console.WriteLine("\n\u001B[36m🔔 Дневник напоминаний (C#)\u001B[0m");
            Console.WriteLine("1. Добавить напоминание");
            Console.WriteLine("2. Показать все напоминания");
            Console.WriteLine("3. Показать предстоящие напоминания");
            Console.WriteLine("4. Поиск напоминаний");
            Console.WriteLine("5. Удалить напоминание");
            Console.WriteLine("6. Отметить как выполненное");
            Console.WriteLine("7. Статистика");
            Console.WriteLine("8. Выход");
            Console.Write("Выберите действие: ");
            string choice = Console.ReadLine();
            switch (choice)
            {
                case "1": AddReminder(); break;
                case "2": ListAll(); break;
                case "3": ListUpcoming(); break;
                case "4": SearchReminders(); break;
                case "5": DeleteReminder(); break;
                case "6": MarkDone(); break;
                case "7": Stats(); break;
                case "8": Console.WriteLine("До свидания!"); return;
                default: Console.WriteLine("\u001B[31mНеверный выбор.\u001B[0m"); break;
            }
        }
    }

    static void Load()
    {
        if (File.Exists(DataFile))
        {
            try
            {
                string json = File.ReadAllText(DataFile);
                reminders = JsonSerializer.Deserialize<List<Reminder>>(json) ?? new List<Reminder>();
            }
            catch { reminders = new List<Reminder>(); }
        }
    }

    static void Save()
    {
        string json = JsonSerializer.Serialize(reminders, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(DataFile, json);
    }

    static void AddReminder()
    {
        Console.Write("Название: ");
        string title = Console.ReadLine();
        Console.Write("Дата (ГГГГ-ММ-ДД): ");
        string date = Console.ReadLine();
        Console.Write("Пробег (км): ");
        int mileage = int.Parse(Console.ReadLine());
        Console.Write("Описание: ");
        string desc = Console.ReadLine();
        Console.Write("Приоритет (низкий/средний/высокий): ");
        string priority = Console.ReadLine().ToLower();
        if (priority != "низкий" && priority != "средний" && priority != "высокий") priority = "средний";
        int id = reminders.Count + 1;
        reminders.Add(new Reminder { Id = id, Title = title, Date = date, Mileage = mileage, Description = desc, Priority = priority, Done = false });
        Save();
        Console.WriteLine($"\u001B[32m✅ Напоминание добавлено (ID: {id})\u001B[0m");
    }

    static void ListAll()
    {
        if (reminders.Count == 0)
        {
            Console.WriteLine("\u001B[33mНет напоминаний.\u001B[0m");
            return;
        }
        Console.WriteLine($"\u001B[36m{"ID",-4} {"Название",-20} {"Дата",-12} {"Пробег",-10} {"Приоритет",-10} {"Статус",-12}\u001B[0m");
        Console.WriteLine(new string('-', 80));
        foreach (var r in reminders)
        {
            string status = r.Done ? "✅ Выполнено" : "⏳ Ожидает";
            string color = r.Priority == "низкий" ? "\u001B[32m" : r.Priority == "средний" ? "\u001B[33m" : "\u001B[31m";
            Console.WriteLine($"{r.Id,-4} {r.Title,-20} {r.Date,-12} {r.Mileage,-10} {color}{r.Priority,-10}\u001B[0m {status,-12}");
        }
    }

    static void ListUpcoming()
    {
        Console.Write("Количество дней (по умолч. 7): ");
        string daysStr = Console.ReadLine();
        int days = string.IsNullOrEmpty(daysStr) ? 7 : int.Parse(daysStr);
        var now = DateTime.Now.Date;
        var upcoming = reminders.Where(r => !r.Done && DateTime.Parse(r.Date) >= now && DateTime.Parse(r.Date) <= now.AddDays(days)).ToList();
        if (upcoming.Count == 0)
        {
            Console.WriteLine($"\u001B[32mНет напоминаний на ближайшие {days} дней.\u001B[0m");
            return;
        }
        Console.WriteLine($"\u001B[36m📅 Напоминания на ближайшие {days} дней:\u001B[0m");
        foreach (var r in upcoming)
        {
            int diff = (DateTime.Parse(r.Date) - now).Days;
            Console.WriteLine($"  {r.Id}: {r.Title} — {r.Date} (осталось {diff} дн)");
        }
    }

    static void SearchReminders()
    {
        Console.Write("Введите ключевое слово: ");
        string keyword = Console.ReadLine().ToLower();
        var results = reminders.Where(r => r.Title.ToLower().Contains(keyword) || r.Description.ToLower().Contains(keyword)).ToList();
        if (results.Count == 0)
            Console.WriteLine("\u001B[33mНичего не найдено.\u001B[0m");
        else
            results.ForEach(r => Console.WriteLine($"{r.Id}: {r.Title} | {r.Date} | {r.Priority} | {(r.Done ? "✅" : "⏳")}"));
    }

    static void DeleteReminder()
    {
        ListAll();
        Console.Write("Введите ID для удаления: ");
        int id = int.Parse(Console.ReadLine());
        var item = reminders.FirstOrDefault(r => r.Id == id);
        if (item != null)
        {
            reminders.Remove(item);
            Save();
            Console.WriteLine("\u001B[32m✅ Напоминание удалено.\u001B[0m");
        }
        else
            Console.WriteLine("\u001B[31m❌ Напоминание не найдено.\u001B[0m");
    }

    static void MarkDone()
    {
        ListAll();
        Console.Write("Введите ID для отметки как выполненное: ");
        int id = int.Parse(Console.ReadLine());
        var item = reminders.FirstOrDefault(r => r.Id == id);
        if (item != null)
        {
            item.Done = true;
            Save();
            Console.WriteLine("\u001B[32m✅ Напоминание отмечено как выполненное.\u001B[0m");
        }
        else
            Console.WriteLine("\u001B[31m❌ Напоминание не найдено.\u001B[0m");
    }

    static void Stats()
    {
        int total = reminders.Count;
        int done = reminders.Count(r => r.Done);
        int pending = total - done;
        var now = DateTime.Now.Date;
        int overdue = reminders.Count(r => !r.Done && DateTime.Parse(r.Date) < now);
        Console.WriteLine("\u001B[36m📊 Статистика:\u001B[0m");
        Console.WriteLine($"  Всего напоминаний: {total}");
        Console.WriteLine($"  Выполнено: {done}");
        Console.WriteLine($"  Ожидает: {pending}");
        Console.WriteLine($"  Просрочено: {overdue}");
    }
}
